# Исследование надежности заемщиков


Во второй части проекта вы выполните шаги 3 и 4. Их вручную проверит ревьюер.
Чтобы вам не пришлось писать код заново для шагов 1 и 2, мы добавили авторские решения в ячейки с кодом. 



## Откройте таблицу и изучите общую информацию о данных

**Задание 1. Импортируйте библиотеку pandas. Считайте данные из csv-файла в датафрейм и сохраните в переменную `data`. Путь к файлу:**

`/datasets/data.csv`

In [1]:
import pandas as pd

try:
    data = pd.read_csv('\\C:\\Users\\79184\\Desktop\\проект 1\\data.csv')
except:
    data = pd.read_csv('https://code.s3.yandex.net/datasets/data.csv')

**Задание 2. Выведите первые 20 строчек датафрейма `data` на экран.**

In [2]:
data.head(20)

children  days_employed  dob_years            education  education_id  \
0          1   -8437.673028         42               высшее             0   
1          1   -4024.803754         36              среднее             1   
2          0   -5623.422610         33              Среднее             1   
3          3   -4124.747207         32              среднее             1   
4          0  340266.072047         53              среднее             1   
5          0    -926.185831         27               высшее             0   
6          0   -2879.202052         43               высшее             0   
7          0    -152.779569         50              СРЕДНЕЕ             1   
8          2   -6929.865299         35               ВЫСШЕЕ             0   
9          0   -2188.756445         41              среднее             1   
10         2   -4171.483647         36               высшее             0   
11         0    -792.701887         40              среднее             1   
12         0            NaN         65              среднее             1   
13         0   -1846.641941         54  неоконченное высшее             2   
14         0   -1844.956182         56               высшее             0   
15         1    -972.364419         26              среднее             1   
16         0   -1719.934226         35              среднее             1   
17         0   -2369.999720         33               высшее             0   
18         0  400281.136913         53              среднее             1   
19         0  -10038.818549         48              СРЕДНЕЕ             1   

       family_status  family_status_id gender income_type  debt  \
0    женат / замужем                 0      F   сотрудник     0   
1    женат / замужем                 0      F   сотрудник     0   
2    женат / замужем                 0      M   сотрудник     0   
3    женат / замужем                 0      M   сотрудник     0   
4   гражданский брак                 1      F   пенсионер     0   
5   гражданский брак                 1      M   компаньон     0   
6    женат / замужем                 0      F   компаньон     0   
7    женат / замужем                 0      M   сотрудник     0   
8   гражданский брак                 1      F   сотрудник     0   
9    женат / замужем                 0      M   сотрудник     0   
10   женат / замужем                 0      M   компаньон     0   
11   женат / замужем                 0      F   сотрудник     0   
12  гражданский брак                 1      M   пенсионер     0   
13   женат / замужем                 0      F   сотрудник     0   
14  гражданский брак                 1      F   компаньон     1   
15   женат / замужем                 0      F   сотрудник     0   
16   женат / замужем                 0      F   сотрудник     0   
17  гражданский брак                 1      M   сотрудник     0   
18    вдовец / вдова                 2      F   пенсионер     0   
19         в разводе                 3      F   сотрудник     0   

     total_income                                 purpose  
0   253875.639453                           покупка жилья  
1   112080.014102                 приобретение автомобиля  
2   145885.952297                           покупка жилья  
3   267628.550329              дополнительное образование  
4   158616.077870                         сыграть свадьбу  
5   255763.565419                           покупка жилья  
6   240525.971920                       операции с жильем  
7   135823.934197                             образование  
8    95856.832424                   на проведение свадьбы  
9   144425.938277                 покупка жилья для семьи  
10  113943.491460                    покупка недвижимости  
11   77069.234271       покупка коммерческой недвижимости  
12            NaN                         сыграть свадьбу  
13  130458.228857                 приобретение автомобиля  
14  165127.911772              покупка жилой недвижимости  
15  116820.904450 

**Задание 3. Выведите основную информацию о датафрейме с помощью метода `info()`.**

In [3]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 21525 entries, 0 to 21524
Data columns (total 12 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   children          21525 non-null  int64  
 1   days_employed     19351 non-null  float64
 2   dob_years         21525 non-null  int64  
 3   education         21525 non-null  object 
 4   education_id      21525 non-null  int64  
 5   family_status     21525 non-null  object 
 6   family_status_id  21525 non-null  int64  
 7   gender            21525 non-null  object 
 8   income_type       21525 non-null  object 
 9   debt              21525 non-null  int64  
 10  total_income      19351 non-null  float64
 11  purpose           21525 non-null  object 
dtypes: float64(2), int64(5), object(5)
memory usage: 2.0+ MB


## Предобработка данных

### Удаление пропусков

**Задание 4. Выведите количество пропущенных значений для каждого столбца. Используйте комбинацию двух методов.**

In [4]:
data.isna().sum()

children               0
days_employed       2174
dob_years              0
education              0
education_id           0
family_status          0
family_status_id       0
gender                 0
income_type            0
debt                   0
total_income        2174
purpose                0
dtype: int64

**Задание 5. В двух столбцах есть пропущенные значения. Один из них — `days_employed`. Пропуски в этом столбце вы обработаете на следующем этапе. Другой столбец с пропущенными значениями — `total_income` — хранит данные о доходах. На сумму дохода сильнее всего влияет тип занятости, поэтому заполнить пропуски в этом столбце нужно медианным значением по каждому типу из столбца `income_type`. Например, у человека с типом занятости `сотрудник` пропуск в столбце `total_income` должен быть заполнен медианным доходом среди всех записей с тем же типом.**

In [5]:
for t in data['income_type'].unique():
    data.loc[(data['income_type'] == t) & (data['total_income'].isna()), 'total_income'] = \
    data.loc[(data['income_type'] == t), 'total_income'].median()

### Обработка аномальных значений

**Задание 6. В данных могут встречаться артефакты (аномалии) — значения, которые не отражают действительность и появились по какой-то ошибке. таким артефактом будет отрицательное количество дней трудового стажа в столбце `days_employed`. Для реальных данных это нормально. Обработайте значения в этом столбце: замените все отрицательные значения положительными с помощью метода `abs()`.**

In [6]:
data['days_employed'] = data['days_employed'].abs()

**Задание 7. Для каждого типа занятости выведите медианное значение трудового стажа `days_employed` в днях.**

In [7]:
data.groupby('income_type')['days_employed'].agg('median')

income_type
безработный        366413.652744
в декрете            3296.759962
госслужащий          2689.368353
компаньон            1547.382223
пенсионер          365213.306266
предприниматель       520.848083
сотрудник            1574.202821
студент               578.751554
Name: days_employed, dtype: float64

У двух типов (безработные и пенсионеры) получатся аномально большие значения. Исправить такие значения сложно, поэтому оставьте их как есть. Тем более этот столбец не понадобится вам для исследования.

**Задание 8. Выведите перечень уникальных значений столбца `children`.**

In [8]:
data['children'].unique()

array([ 1,  0,  3,  2, -1,  4, 20,  5], dtype=int64)

**Задание 9. В столбце `children` есть два аномальных значения. Удалите строки, в которых встречаются такие аномальные значения из датафрейма `data`.**

In [9]:
data = data[(data['children'] != -1) & (data['children'] != 20)]

**Задание 10. Ещё раз выведите перечень уникальных значений столбца `children`, чтобы убедиться, что артефакты удалены.**

In [10]:
data['children'].unique()

array([1, 0, 3, 2, 4, 5], dtype=int64)

### Удаление пропусков (продолжение)

**Задание 11. Заполните пропуски в столбце `days_employed` медианными значениями по каждого типа занятости `income_type`.**

In [11]:
for t in data['income_type'].unique():
    data.loc[(data['income_type'] == t) & (data['days_employed'].isna()), 'days_employed'] = \
    data.loc[(data['income_type'] == t), 'days_employed'].median()

**Задание 12. Убедитесь, что все пропуски заполнены. Проверьте себя и ещё раз выведите количество пропущенных значений для каждого столбца с помощью двух методов.**

In [12]:
data.isna().sum()

children            0
days_employed       0
dob_years           0
education           0
education_id        0
family_status       0
family_status_id    0
gender              0
income_type         0
debt                0
total_income        0
purpose             0
dtype: int64

### Изменение типов данных

**Задание 13. Замените вещественный тип данных в столбце `total_income` на целочисленный с помощью метода `astype()`.**

In [13]:
data['total_income'] = data['total_income'].astype(int)

### Обработка дубликатов

**Задание 14. Обработайте неявные дубликаты в столбце `education`. В этом столбце есть одни и те же значения, но записанные по-разному: с использованием заглавных и строчных букв. Приведите их к нижнему регистру. Проверьте остальные столбцы.**

In [14]:
data['education'] = data['education'].str.lower()

**Задание 15. Выведите на экран количество строк-дубликатов в данных. Если такие строки присутствуют, удалите их.**

In [15]:
data.duplicated().sum()

71

In [16]:
data = data.drop_duplicates()

### Категоризация данных

**Задание 16. На основании диапазонов, указанных ниже, создайте в датафрейме `data` столбец `total_income_category` с категориями:**

- 0–30000 — `'E'`;
- 30001–50000 — `'D'`;
- 50001–200000 — `'C'`;
- 200001–1000000 — `'B'`;
- 1000001 и выше — `'A'`.


**Например, кредитополучателю с доходом 25000 нужно назначить категорию `'E'`, а клиенту, получающему 235000, — `'B'`. Используйте собственную функцию с именем `categorize_income()` и метод `apply()`.**

In [17]:
def categorize_income(income):
    try:
        if 0 <= income <= 30000:
            return 'E'
        elif 30001 <= income <= 50000:
            return 'D'
        elif 50001 <= income <= 200000:
            return 'C'
        elif 200001 <= income <= 1000000:
            return 'B'
        elif income >= 1000001:
            return 'A'
    except:
        pass

In [18]:
data['total_income_category'] = data['total_income'].apply(categorize_income)

**Задание 17. Выведите на экран перечень уникальных целей взятия кредита из столбца `purpose`.**

In [19]:
data['purpose'].unique()

array(['покупка жилья', 'приобретение автомобиля',
       'дополнительное образование', 'сыграть свадьбу',
       'операции с жильем', 'образование', 'на проведение свадьбы',
       'покупка жилья для семьи', 'покупка недвижимости',
       'покупка коммерческой недвижимости', 'покупка жилой недвижимости',
       'строительство собственной недвижимости', 'недвижимость',
       'строительство недвижимости', 'на покупку подержанного автомобиля',
       'на покупку своего автомобиля',
       'операции с коммерческой недвижимостью',
       'строительство жилой недвижимости', 'жилье',
       'операции со своей недвижимостью', 'автомобили',
       'заняться образованием', 'сделка с подержанным автомобилем',
       'получение образования', 'автомобиль', 'свадьба',
       'получение дополнительного образования', 'покупка своего жилья',
       'операции с недвижимостью', 'получение высшего образования',
       'свой автомобиль', 'сделка с автомобилем',
       'профильное образование', 'высшее об

**Задание 18. Создайте функцию, которая на основании данных из столбца `purpose` сформирует новый столбец `purpose_category`, в который войдут следующие категории:**

- `'операции с автомобилем'`,
- `'операции с недвижимостью'`,
- `'проведение свадьбы'`,
- `'получение образования'`.

**Например, если в столбце `purpose` находится подстрока `'на покупку автомобиля'`, то в столбце `purpose_category` должна появиться строка `'операции с автомобилем'`.**

**Используйте собственную функцию с именем `categorize_purpose()` и метод `apply()`. Изучите данные в столбце `purpose` и определите, какие подстроки помогут вам правильно определить категорию.**

In [20]:
def categorize_purpose(row):
    try:
        if 'автом' in row:
            return 'операции с автомобилем'
        elif 'жил' in row or 'недвиж' in row:
            return 'операции с недвижимостью'
        elif 'свад' in row:
            return 'проведение свадьбы'
        elif 'образов' in row:
            return 'получение образования'
    except:
        return 'нет категории'

In [21]:
data['purpose_category'] = data['purpose'].apply(categorize_purpose)

### Шаг 3. Исследуйте данные и ответьте на вопросы

#### 3.1 Есть ли зависимость между количеством детей и возвратом кредита в срок?

In [22]:
data['children'].unique() # вывод уникальных значений количества детей

array([1, 0, 3, 2, 4, 5], dtype=int64)

In [23]:
data['debt'].unique() # вывод уникальных значений обозначения задолженности (1 = есть, 0 = нет)

array([0, 1], dtype=int64)

In [24]:
data['debt'].value_counts() # количество должников и заемщиков без долга

0    19599
1     1732
Name: debt, dtype: int64

In [25]:
((data['debt'] == 1) / data['debt'].count()).sum() * 100

8.119638085415593

In [26]:
data_children = data.groupby('children').agg({'debt': ['mean', 'count', 'sum']}) #групируем данные по количеству детей и наличию задолженности
data_children.sort_values(by='children')

debt             
              mean  count   sum
children                       
0         0.075438  14091  1063
1         0.092346   4808   444
2         0.094542   2052   194
3         0.081818    330    27
4         0.097561     41     4
5         0.000000      9     0

In [27]:
def have_a_child(children): # функция для выявления наличия/отсутствия детей
    if children == 0:
        return 0
    return 1

In [28]:
data['have_a_child'] = data['children'].apply(have_a_child) # добавление столбца have_a_child показвающего наличие(1) и отсутствие (0) детей

In [29]:
data_children = data.groupby('have_a_child').agg({'debt': ['mean', 'count', 'sum']}) # групируем данные по наличию детей и задолжности
display(data_children)

debt             
                  mean  count   sum
have_a_child                       
0             0.075438  14091  1063
1             0.092403   7240   669

**Вывод:**  В ходе исследования получены следущие результаты: общее количество должников 8,12% (1732 случая). В разрезе должников по количеству детей наибольший процент невозврата заемных средств составляет 9,76% у должников с четырьмя детьми. На втором месте (9,45%) находятся должники с двумя детьми, третье место у должников с 1 ребенком (9,23%). Наибольшее количество заемщиков у должников без детей, однако процент невозврата меньше чем у должников с детьми на 1,7 процента (7,54% против 9,24%). В категории заемщиков с пятью детьми случаев невозврата заемных средств нет.  Однако обьективно оценить влияние 3, 4 и 5 детей на возврат кредита в данном иследовании не предстваляется возможным т.к. обьема данных в этих категориях недостаточно (330, 41 и 9 против нескольких тысяч заемщиков в других категориях). Не принимая во внимание данные группы, все равно сохранится взаимосвязь между заемщиками с детьми и возратом займа в срок. Значит, заемщики без детей, являются более добросовесными клиентами.

#### 3.2 Есть ли зависимость между семейным положением и возвратом кредита в срок?

In [30]:
data['family_status'].unique() # уникальные значения семейного положения

array(['женат / замужем', 'гражданский брак', 'вдовец / вдова',
       'в разводе', 'Не женат / не замужем'], dtype=object)

In [31]:
data_family = data.groupby('family_status').agg({'debt': ['mean', 'count', 'sum']}) #групируем данные по семейному положению и наличию задолженности
display(data_family)

debt            
                           mean  count  sum
family_status                              
Не женат / не замужем  0.097639   2796  273
в разводе              0.070648   1189   84
вдовец / вдова         0.066246    951   63
гражданский брак       0.093130   4134  385
женат / замужем        0.075606  12261  927

**Вывод:** Наибольший процент невозврата заемных средств у не женатых / не замужних должников (9,76%) и должников находящихся в гражданском браке (9,31%). Женатые / замужние должники находятся на третьем месте (7,56%) и, также, включают наибольшее количество случаев невозврата. Должники в разводе и вдовцы / вдовы занимают 4 и 5 место соответствено (7,06 % и 6,62%). На основании данных можно сделать вывод, что женатые / замужние должники, а также категории должников, бывшие в прошлом в данном статусе более надежные заемщики чем не женатые / не замужние должники и должники в гражданском браке.

#### 3.3 Есть ли зависимость между уровнем дохода и возвратом кредита в срок?

In [32]:
data['total_income_category'].unique() # уникальные значения категорий дохода

array(['B', 'C', 'D', 'E', 'A'], dtype=object)

In [33]:
# data.head(10)

In [34]:
data_income_category = data.groupby('total_income_category').agg({'debt': ['mean', 'count', 'sum']}) #групируем данные по категории дохода и наличию задолженности
display(data_income_category)

debt             
                           mean  count   sum
total_income_category                       
A                      0.080000     25     2
B                      0.070602   5014   354
C                      0.084982  15921  1353
D                      0.060172    349    21
E                      0.090909     22     2

**Вывод:** Наибольший процент невозврата (9,09 %)  у должников с низким доходом (Е категория), однако и количество заемщиков данной категории всего 2. На втором месте находятся должники из самой крупной категории со средним уровнем доходов (С категория) (8,85 %). Высокий доход не является показателем надежности должника (А категория, 8 %), но и количество должников в данной категории всего 2. Должники с доходом выше среднего (В категория) не возвращают заем в 7,06 % случаев. Самыми надежными заемщиками являются должники с доходом ниже среднего (D категория, 6,02 %). 

#### 3.4 Как разные цели кредита влияют на его возврат в срок?

In [35]:
data['purpose_category'].unique() # уникальные значения категорий целей кредита

array(['операции с недвижимостью', 'операции с автомобилем',
       'получение образования', 'проведение свадьбы'], dtype=object)

In [36]:
data_purpose_cat = data.groupby('purpose_category').agg({'debt': ['mean', 'count', 'sum']}) #групируем данные по целям кредита и наличию задолженности
display(data_purpose_cat)

debt            
                              mean  count  sum
purpose_category                              
операции с автомобилем    0.093480   4279  400
операции с недвижимостью  0.072551  10751  780
получение образования     0.092528   3988  369
проведение свадьбы        0.079118   2313  183

**Вывод:** Должники берущие займы на операции с автомобилем и на образование являются наиболее ненадежными заемщиками (9,35 % и 9,25%). Самыми ответственными являются должники берущие заемные средства на недвижимость (7,26 %). На третьем месте должники должники берущие займы на проведение свадьбы (7,91 %).

#### 3.5 Приведите возможные причины появления пропусков в исходных данных.

*Ответ:*  Пропуски в исходных данных были в 2х столбцах: days_employed — общий трудовой стаж в днях и total_income — ежемесячный доход. Причинами пропусков могут быть как ошибки ввода данных, так и сокрытитие информации. Также в days_employed пропуски могут быть связаны с тем, что у некоторых категорий трудовой стаж может отсутствовать (например студенты), а в столбце total_income может отсутствовать доход (например безработный). 

#### 3.6 Объясните, почему заполнить пропуски медианным значением — лучшее решение для количественных переменных.

*Ответ:* Для количественных переменных,  в нашем исследовании, медиана является лучшим решением для заполнения пропусков т.к. медиана является серединным значением из столбцов, в которых заполнялись пропуски. Заполнение пропусков средним арифметическим в данном исследовании неэффективно (к примеру, пропуски заполнялись в столбце total_income — "ежемесячный доход", и, соответственно, применение среднего арифметического к столбцу с большим разбросом значений  даже в одной категории значительно исказило бы результаты анализа (например в категории 'B' - ежемесячный доход от 200001 до 1000000" т.е. может быть, что один или несколько заемщиков получают около 1 млн., а остальные заемщики получают намного более скромную суммму превышающую 200 тыс.)).

### Шаг 4: общий вывод.

**Вывод** 


Целью данного исследования является выявление взаимосвязей при выдаче заемных денежных средств различным категориям заемщиков.


В данном исследовании были поставлены следующие вопросы:
+ **Вопрос 1** Есть ли зависимость между количеством детей и возвратом кредита в срок?
+ **Вопрос 2** Есть ли зависимость между семейным положением и возвратом кредита в срок?
+ **Вопрос 3** Есть ли зависимость между уровнем дохода и возвратом кредита в срок?
+ **Вопрос 4** Как разные цели кредита влияют на его возврат в срок?



Для полного ответа на **Вопрос 1** мы не имеем достаточного количества данных о должниках с 3, 4 и 5 детьми, однако,  даже не принимая во внимание данные группы, все равно сохраняется взаимосвязь между заемщиками с детьми и возратом займа в срок, **значит наличие детей влияет на возврат кредита в срок**, что подтверждается результатами исследования:
+ Доля должников с детьми против доли должников без детей **9,24%** против **7,54%** 
+ В категории заемщиков с 1 ребенком доля должников по отношению ко всей данной категории сотавляет 9,23%
+ В категории заемщиков с 2 детьми доля должников по отношению ко всей данной категории сотавляет 9,45%



Отвечая на **Вопрос 2** , можно с уверенностью утверждать, что __заемщики состоящие в браке, а также заемщики состоявшие в браке в прошлом более надежные заемщики__, подтверждая данный вывод следующими результатами исследования:

+ Доля невозврата заемных средств у не женатых / не замужних должников **9,76%** и должников находящихся в гражданском браке **9,31%**
+ Заемщики состоящие в браке, а также заемщики состоявшие в браке в прошлом (должники в разводе и вдовы / вдовцы) составляют **7,56%**, **7,06%** и **6,62%**


Данные, анализируемые по категориям при ответе на **Вопрос 3**, не являются достаточными для полного ответа на данный вопрос в связи с маленьким обьемом информации по категориям заемщиков с высоким доходом(А-25 заемщиков), доходом ниже среднего(D-349 заемщиков) и низким доходом(Е-22 заемщика). Однако,если не принимать во внимание данные категории, __доля должников снижается при увеличении уровня дохода заемщика__,а именно:

+ Заемщики получающие от 200 тыс. до 1 млн. более добросовесные плательщики займа чем заемщики с заработной платой от 500 тыс. до 200 тыс. (**7,06%** против **8,49%**)



Анализируя взаимосвясь между целями кредита и его возвратом в срок (**Вопрос 4**), __минимальный процент невозврата кредита в установленный срок наблюдается среди операций с недвижимостью (7,26%)__, по остальным категориям мы наблюдаем следующие результаты:
+ **7,91%** должников берущих займ на проведение свадьбы
+ **9,25%** должников берущих займ на получение образования
+ **9,35%** должников берущих займ на операции с автомобилем


__Значит, заемщики берущие займ на операции с автомобилем или на получение образования примерно на 2% чаще не отдают кредит в срок в отличие от остальных категорий__


*Однако из соотношения данных, представленных для анализа, нельзя быть окончательно уверенным, что при равных обьемах заемщиков результаты остались бы на том же уровне т.к. больше половины заемщиков (10751 из 21331 заемщиков) приходится на операции с недвижимостью*



**Общие рекомендации**

1. Результаты анализа, полученые в ходе исследования Вопроса 2 можно использовать в работе.
2. По 1, 3, 4 вопросу необходимо собрать выборку с равным соотношением данных по рассматриваемым категориям в целях получения достоверных показателей.